source/agora/utils/InetUtils.d

```diff
@@ -0,0 +1,224 @@
+                                buffer.ptr,
+                                buffer.length,
+                                null,
+                                0,
+                                NI_NUMERICHOST);
+                if (name_info_res)
+                {
+                    log.error("error happened during a call to getnameinfo, name_info_res code:", name_info_res);
+                    continue;
+                }
+                string ip = buffer[0 .. strlen(buffer.ptr)].idup();
+                ips ~= ip;
+            }
+        }
+
+        return ips;
+    }
+}
+
+version(OSX)
+{
+    static if (__VERSION__ < 2095)
+    {
+        extern (C):
+        nothrow:
+        @nogc:
+
+        struct ifaddrs
+        {
+            /// Next item in the list
+            ifaddrs* ifa_next;
+            /// Name of the interface
+            char* ifa_name;
+            /// Flags from SIOCGIFFLAGS
+            uint ifa_flags;
+            /// Address of interface
+            sockaddr* ifa_addr;
+            /// Netmask of interface
+            sockaddr* ifa_netmask;
+            /// Point-to-point destination addresss
+            sockaddr* if_dstaddr;
+            /// Address specific data
+            void* ifa_data;
+        }
+
+        /// Returns: linked list of ifaddrs structures describing interfaces
+        int getifaddrs(ifaddrs** );
+        /// Frees the linked list returned by getifaddrs
+        void freeifaddrs(ifaddrs* );
+    }
+    else
+    {
+        import core.sys.darwin.ifaddrs;
+    }
+}
+
+struct InetUtils
+{
+version(OSX)
+{
+    import core.sys.posix.netdb;
+    import core.sys.posix.netinet.in_;
+    import core.sys.posix.sys.socket;
+
+    mixin AddGetAllIPs;
+}
+version(linux)
+{
+    import core.sys.linux.ifaddrs;
+    import core.sys.posix.netdb;
+
+    mixin AddGetAllIPs;
+}
+version (Windows)
+{
+    import std.socket;
+
+    import core.sys.windows.iphlpapi;
+    import core.sys.windows.iptypes;
+    import core.sys.windows.windef;
+    import core.sys.windows.winsock2;
+    import core.stdc.stdlib: malloc, free;
+    import core.stdc.string: strlen;
+
+    public static string[] getAllIPs()
+    {
+        string[] ips;
+        PIP_ADAPTER_INFO adapter_info_head;
+        PIP_ADAPTER_INFO adapter_info;
+        DWORD ret_adapters_info;
+        ULONG buff_length = IP_ADAPTER_INFO.sizeof;
+        adapter_info_head = cast(IP_ADAPTER_INFO *) malloc(IP_ADAPTER_INFO.sizeof);
+        if (adapter_info_head == NULL)
+        {
+            log.error("Error allocating memory needed to call GetAdaptersinfo - 1");
+            return [];
+        }
+        scope(exit) free(adapter_info_head);
+        // find out the real size we need to allocate
+        if (GetAdaptersInfo(adapter_info_head, &buff_length) == ERROR_BUFFER_OVERFLOW)
+        {
+            free(adapter_info_head);
+            adapter_info_head = cast(IP_ADAPTER_INFO *) malloc(buff_length);
```

Comment:
realloc ?

Reply:
Realloc can be more more costly than malloc, in the case where there isn't enough space after the buffer previously allocated by malloc. In that case realloc would not only call malloc, but would unnecessarily copy all the contents from the previous memory location.

Reply:
Good point, lets leave it as is.